protocol/consistencychecks.go

```diff
@@ -0,0 +1,265 @@
+	return &ConsistencyChecks{
+		CurrentEpoch: 0,
+		SavedSTR:     savedSTR,
+		isUseTBs:     isUseTBs,
+		signKey:      signKey,
+	}
+}
+
+// Verify verifies the directory's response for a request.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response,
+// and finally, it verifies the returned proof type
+// according to the request type.
+// This will also update the consistency state if the check is passed.
+func (cc *ConsistencyChecks) Verify(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var verificationResult error
+	var proofType int
+
+	switch msg.DirectoryResponse.(type) {
+	case *DirectoryProof:
+		proofType, verificationResult = cc.verifyDirectoryProof(
+			msg.DirectoryResponse.(*DirectoryProof), uname, key)
+	case *DirectoryProofs:
+		proofType, verificationResult = cc.verifyDirectoryProofs(
+			msg.DirectoryResponse.(*DirectoryProofs), uname, key)
+	}
+
+	if verificationResult != nil {
+		return verificationResult.(ErrorCode)
+	}
+
+	switch requestType {
+	case RegistrationType:
+		if msg.Error == ErrorNameExisted &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		} else if msg.Error == Success &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		}
+		return cc.verifyProofTypeWithTB(msg.Error, proofType)
+
+	case KeyLookupType:
+		if msg.Error == ErrorNameNotFound &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		} else if msg.Error == Success &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		}
+		return cc.verifyProofTypeWithTB(msg.Error, proofType)
+
+	case MonitoringType:
+		if proofType == proofOfAbsence {
+			return ErrorBadProofType
+		}
+		return PassedWithAProofOfInclusion
+
+	case KeyLookupInEpochType:
+		// TODO: fix me
+
+	}
+
+	// this return should be unreachable
+	// unless the developers mistakenly passed an invalid request type.
+	return ErrorMalformedClientMessage
+}
+
+func (cc *ConsistencyChecks) verifyDirectoryProof(df *DirectoryProof,
+	uname string, key []byte) (proofType int, verificationResult error) {
+
+	verificationResult = nil
+	proofType = invalidProof
+
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if verificationResult = cc.verifySTR(str); verificationResult != nil {
+		return
+	}
+
+	if proofType, verificationResult = cc.verifyAuthPath(uname, key,
+		ap, str); verificationResult != nil {
+		return
+	}
+
+	if verificationResult = cc.verifyReturnedPromise(tb, str, ap); verificationResult != nil {
+		return
+	}
+
+	if verificationResult = cc.verifyFulfilledPromise(str, ap); verificationResult != nil {
+		return
+	}
+
+	// re-assign new state for the client
+	if str.Epoch > cc.CurrentEpoch {
```

Comment:
I know that `verifySTR()` checks that `str.Epoch == cc.CurrentEpoch+1` but I think for consistency, I would prefer to repeat that check here.